scalyr_agent/builtin_monitors/linux_process_metrics.py

```diff
@@ -619,163 +741,336 @@ class ProcessMonitor(ScalyrMonitor):
+        for pid, process_metrics in self.__metrics_history.items():
+            for _metric, _metric_values in process_metrics.items():
+                if not _metric.is_cumulative:
+                    self.__aggregated_metrics[_metric] = 0
 
-        If the process is no longer running, then attempts to match a new one.
+    def _calculate_aggregated_metrics(self, running_pids):
+        """
+        Calculates the aggregated metric values based on the current running processes
+        and the historical metric record
+        @param running_pids: list of running process ids
+        @type running_pids: list
         """
-        if self.__pid is not None and not self.__is_running():
-            self.__set_process(None)
 
-        if self.__pid is None:
-            self.__set_process(self.__select_process())
+        # using the historical values, calculate the aggregate
+        # there are two kinds of metrics:
+        # a) cumulative metrics - only the delta of the last 2 recorded values is used (eg cpu cycles)
+        # b) absolute metrics - the last absolute value is used
+
+        running_pids_set = set(running_pids)
+
+        for pid, process_metrics in self.__metrics_history.items():
+            for _metric, _metric_values in process_metrics.items():
+                if not self.__aggregated_metrics.get(_metric):
+                    self.__aggregated_metrics[_metric] = 0
+                if _metric.is_cumulative:
+                    if pid in running_pids_set:
+                        if len(_metric_values) > 1:
+                            # only report the cumulative metrics for more than one sample
+                            self.__aggregated_metrics[_metric] += (_metric_values[-1] - _metric_values[-2])
+                        else:
+                            self.__aggregated_metrics[_metric] = 0
+                else:
+                    if pid in running_pids_set:
+                        # absolute metric - accumulate the last reported value
+                        self.__aggregated_metrics[_metric] += _metric_values[-1]
+
+    def _remove_dead_processes(self, running_pids):
+        # once the _calculate_aggregated_metrics has calculated the running total of the
+        # metrics for the current epoch, remove the entries for the process id in the history
+        # NOTE: the removal of the contributions (if applicable) should already be done so
+        # removing the entry from the history is safe.
+
+        all_pids = self.__metrics_history.keys()
+        for _pid_to_remove in list(set(all_pids) - set(running_pids)):
+            # for all the absolute metrics, decrease the count that the dead processes accounted for
+            del self.__metrics_history[_pid_to_remove]
+
+        # if no processes are running, there is no reason to report the running metric data
+        if not running_pids:
+            self.__aggregated_metrics = {}
+
+    def get_pids_from_ps(self, subprocesses=False, ppid=None):
```

Comment:
I think this is getting too complicated and would be better if broke it out into a smaller abstraction.

Call it `ProcessList`.  When it is instantiated, it invokes `ps -axo pid,ppid,command` to get the complete list in one go.  It saves that information in its state variables.

Then `ProcessList` should have the following method:
  * `get_matches_commandline(input_string)`
  * `get_child_processes(pid)`

You could even then define a final method that uses the first two:
  * get_matches_commandline_and_children(input_string)

You can test this abstraction separately.

Reply:
great idea